components/cop_datatype/def/field_type.rs

```diff
@@ -0,0 +1,321 @@
+// Copyright 2018 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::fmt;
+
+use tipb::expression::FieldType;
+use tipb::schema::ColumnInfo;
+
+use num_traits::FromPrimitive;
+
+/// Valid values of `tipb::expression::FieldType::tp` and `tipb::schema::ColumnInfo::tp`.
+///
+/// `FieldType` is the field type of a column defined by schema.
+///
+/// `ColumnInfo` describes a column. It contains `FieldType` and some other column specific
+/// information. However for historical reasons, fields in `FieldType` (for example, `tp`)
+/// are flattened into `ColumnInfo`. Semantically these fields are identical.
+///
+/// Please refer to `mysql/type.go` in TiDB.
+#[derive(Primitive, PartialEq, Debug, Clone, Copy)]
+pub enum FieldTypeTp {
+    Unspecified = 0, // Default
+    Tiny = 1,
+    Short = 2,
+    Long = 3,
+    Float = 4,
+    Double = 5,
+    Null = 6,
+    Timestamp = 7,
+    LongLong = 8,
+    Int24 = 9,
+    Date = 10,
+    Duration = 11,
+    DateTime = 12,
+    Year = 13,
+    NewDate = 14,
+    VarChar = 15,
+    Bit = 16,
+    JSON = 0xf5,
```

Comment:
Are the numbers like this defined by `TiDB`? 
I think you'd better add comments to those places which may cause confusion.

Reply:
There are already comments describing it. Please see Line 21 - Line 29.